### DISCUSSION TITLE
`airflow kubernetes` sub-command is gone

### DISCUSSION THREAD

**Discussion by slmg** [apache/airflow#35941] [General] [answered]
### Apache Airflow version

2.7.3

### What happened

I tried to run `airflow kubernetes generate-dag-yaml` locally, as documented in [core-concepts/executor/kubernetes.html#fault-tolerance](https://airflow.apache.org/docs/apache-airflow/2.7.3/core-concepts/executor/kubernetes.html#fault-tolerance), but then realised the sub-command `kubernetes` disapeared (it used to work fine on 2.6.3, the prev version I used).

### How to reproduce

```sh
> docker container run --rm apache/airflow:2.7.3-python3.10 airflow kubernetes
airflow command error: argument GROUP_OR_COMMAND: invalid choice: 'kubernetes' (choose from 'cheat-sheet', 'config', 'connections', 'dag-processor', 'dags', 'db', 'info', 'jobs', 'kerberos', 'plugins', 'pools', 'providers', 'roles', 'rotate-fernet-key', 'scheduler', 'standalone', 'sync-perm', 'tasks', 'triggerer', 'users', 'variables', 'version', 'webserver'), see help above.
Usage: airflow [-h] GROUP_OR_COMMAND ...

Positional Arguments:
  GROUP_OR_COMMAND

    Groups
      config         View configuration
      connections    Manage connections
      dags           Manage DAGs
      db             Database operations
      jobs           Manage jobs
      pools          Manage pools
      providers      Display providers
      roles          Manage roles
      tasks          Manage tasks
      users          Manage users
      variables      Manage variables

    Commands:
      cheat-sheet    Display cheat sheet
      dag-processor  Start a standalone Dag Processor instance
      info           Show information about current Airflow and environment
      kerberos       Start a kerberos ticket renewer
      plugins        Dump information about loaded plugins
      rotate-fernet-key
                     Rotate encrypted connection credentials and variables
      scheduler      Start a scheduler instance
      standalone     Run an all-in-one copy of Airflow
      sync-perm      Update permissions for existing roles and optionally DAGs
      triggerer      Start a triggerer instance
      version        Show the version
      webserver      Start a Airflow webserver instance

Options:
  -h, --help         show this help message and exit
```

### Operating System

Linux

### Versions of Apache Airflow Providers

```sh
> pip list | grep -i cncf
apache-airflow-providers-cncf-kubernetes 7.8.0
```

### Deployment

Official Apache Airflow Helm Chart

### Anything else

Likely related to the move of `cncf-kubernetes` off the main codebase that occured between 2.6.3 and 2.7.3.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
You need to have kubernetes provider installed and kubernetes executor configured. See release notes about moving kubernetes executor to provider.

**Reply by potiuk** [reply]
setting KubernetesExecutor as executor is the crucial part since you have provider. converted it to discussion.

**Reply by slmg** [reply]
Hi @potiuk , thank you for your response. I got this to work following your advice.

```sh
docker container run --rm apache/airflow:2.7.3-python3.10 bash -c "AIRFLOW__CORE__EXECUTOR=KubernetesExecutor airflow kubernetes"
```

While I'm happy a workaround exists, as we code from within a [devcontainer](https://code.visualstudio.com/docs/devcontainers/containers) we have the LocalExecutor set. Switching the executor just to `generate-dag-yaml` feels a bit hacky. Inspecting rendered pod templates is a BAU step to perform in dev. In your opinion, would it be better to either / both
* dynamically load sub-commands for a given provider, similar to docker / kubectl behaviour
* render tasks pod templates completely differently?

**Reply by potiuk** [reply, marked as answer]
> dynamically load sub-commands for a given provider, similar to docker / kubectl behaviour

No - that will be very slow because you would have to import a lot of librarries so all the CLIs will get visible slow-down. But if you feel like you can contribute a PR to make the CLI fast (including autocompletion) without it being slow, you are of course welcome. It would likely require quite a deep rewrite of how argparsing is done in Airflow and likely use a different mechanism, but it is posible.

> While I'm happy a workaround exists

It's not a workaround. It's just how it is supposed to work - this CLI is contrbuted by KubernetesExecutor. Your past workflow is broken indeed - but it was working like that accidentally. Celery for example always had celery command working only when Celery Executor was working - so the big change implemented was to make it consistent and not show neither kubernetes nor celery command at all when the executors were not configured.

I suggest you add a simple script that sets the variable and use it to generate your yaml with Airflow CLI. That seems to be a simple solution that is workable for you. Consider airflow CLI as something of an API that you can call using your own wrappers. I guess your developers do not use the CLI for much more than that.
